pages/content/amp-dev/documentation/guides-and-tutorials/develop/custom-javascript-tutorial.md

```diff
@@ -1,231 +1,284 @@
+## Where's my JavaScript?
 
-Open the newly created `script.js` file and add `console.log("amp-script here")`. Reload the page and open the [DevTools console](https://developers.google.com/web/tools/chrome-devtools/) to verify it successfully logged "amp-script here".
+`<amp-script>` runs your JavaScript in a Web Worker. Web Workers can't access the DOM directly, so `<amp-script>` gives the worker access to a virtual copy of the DOM, which it keeps in sync with the real DOM. `<amp-script>` provides emulations of many common DOM APIs, almost all of which you can use in your JavaScript in the usual way.
 
-Imported script logic from the `amp-script` `src` attribute is debugged in the console, same as JavaScript inside non-AMP pages.
+If at any point you need to debug your script, you can set breakpoints in JavaScript in a Web Worker in the same way you do with any JavaScript. You just need to know where to find it. 
 
-# Add logic
+In Chrome DevTools, open the "Sources" tab. At the bottom you will see a long hexadecimal string like the one shown below. Expand that, then expand the "no domain" area, and you'll see your script:
 
-Now that we’ve confirmed our `script.js` file is being injected correctly, let’s add functionality!
+{{ image('/static/img/docs/tutorials/custom-javascript-tutorial/script-in-sources.png', 303, 277, layout='intrinsic', alt='amp-script JavaScript in DevTools Sources panel', align='center' ) }}
 
-HTML elements within the `<amp-script>` tag are now accessible via standard DOM methods from within `script.js`, minus [a few small caveats](custom-javascript.md?format=websites#api-restrictions).
+# Adding our JavaScript
 
-## Declare elements
+Now that we know that our `<amp-script>` is working, let's write some JavaScript!
 
-Our password checker needs to capture the password input box, declare it on `script.js`.
+The first thing we want to do is grab the DOM elements we'll be working with and stash those in globals. Our code will use the password input, the submit button, and the area that shows the password rules. Add these three declarations to `script.js`:
 
 ```js
 const passwordBox = document.getElementById("passwordBox");
+const submitButton = document.getElementById("submitButton");
+const rulesArea = document.getElementById("rules");
 ```
 
-Like non-AMP pages, we can test if it was captured correctly by logging elements into the console.
+Notice that we're able to use regular DOM API methods like `getElementById()`. Although our code runs in a worker, and workers lack direct access to the DOM, `<amp-script>` provides a virtual copy of the DOM and emulates some common APIs, listed [here](https://github.com/ampproject/worker-dom/blob/master/web_compat_table.md). These APIs give us enough tools to cover most use cases. But it's important to note that only a subset of the DOM API is supported. Otherwise, the JavaScript included with `<amp-script>` would be enormous, negating AMP's performance benefits!
```

Comment:
Some APIs we would have trouble recreating, and of course some we cannot do synchronously. But thats only true for a handful. There's far more we don't include simply because it would be a large project and a massive JS hit.

You could imagine of course more ways to do it... like importing just the JS needed for certain APIs :)

Reply:
sgtm if that's the official reason.

Reply:
I don't know if we have an official reason, but it's what Kris etc. have told me :)

Something else to write up in the FAQ, and to confirm with others on the team.